wandb/sdk/wandb_run.py

```diff
@@ -1417,39 +1417,70 @@ def _datatypes_callback(self, fname: str) -> None:
         files: FilesDict = dict(files=[(GlobStr(glob.escape(fname)), "now")])
         self._backend.interface.publish_files(files)
 
+    def _get_nested_value(self, d: Dict[str, Any], keys: List[str]) -> Any:
+        if len(keys) == 1:
+            return d[keys[0]]
+        else:
+            return self._get_nested_value(d[keys[0]], keys[1:])
+
+    def _set_nested_value(
+        self, d: Dict[str, Any], keys: List[str], value: Any, split_table: bool
+    ) -> None:
+        if len(keys) == 1:
+            if split_table:
+                d[f"Custom Chart Tables/{keys[0]}_table"] = value
+            else:
+                d[keys[0] + "_table"] = value
+            d.pop(keys[0])
+        else:
+            self._set_nested_value(d[keys[0]], keys[1:], value, split_table)
+
     def _visualization_hack(self, row: Dict[str, Any]) -> Dict[str, Any]:
-        # TODO(jhr): move visualize hack somewhere else
         chart_keys = set()
         split_table_set = set()
-        for k in row:
-            if isinstance(row[k], Visualize):
-                key = row[k].get_config_key(k)
-                value = row[k].get_config_value(k)
-                row[k] = row[k]._data
-                self._config_callback(val=value, key=key)
-            elif isinstance(row[k], CustomChart):
-                chart_keys.add(k)
-                key = row[k].get_config_key(k)
-                if row[k]._split_table:
-                    value = row[k].get_config_value(
-                        "Vega2", row[k].user_query(f"Custom Chart Tables/{k}_table")
+
+        def transform(key: str, value: Any) -> Any:
+            if isinstance(value, dict):
+                return {k: transform(f"{key}®{k}", v) for k, v in value.items()}
+            if isinstance(value, Visualize):
+                config_key = value.get_config_key(key)
+                config_value = value.get_config_value(key)
+                value = value._data
+                self._config_callback(val=config_value, key=config_key)
+                return value
+            elif isinstance(value, CustomChart):
+                formatted_key = key.replace("®", ".")
```

Comment:
lolwut? :))

Reply:
you'd probably want to use a tuple to keep track of the composite key and when the time comes, join it with dots, it's risky to rely on a char for this